packages/realm-react/src/cachedCollection.ts

```diff
@@ -0,0 +1,165 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+import Realm from "realm";
+
+const numericRegEx = /^-?\d+$/;
+
+function getCacheKey(id: string) {
+  return `${id}`;
+}
+
+/** 
+  * Arguments object for {@link cachedCollection}.
+  */
+type CachedCollectionArgs<T> = {
+  /**
+   * The {@link Realm.Collection} to proxy
+   */
+  collection: Realm.Collection<T>;
+  /**
+   * Callback which is called whenever an object in the collection changes
+   */
+  updateCallback: () => void;
+  /**
+   * Optional Map to be used as the cache. This is used to allow a `sorted` or `filtered`
+   * (derived) version of the collection to reuse the same cache, preventing excess new object
+   * references being created.
+   */
+  collectionCache?: Map<string, T>;
+  /**
+   * Optional flag specifying that this is a derived (`sorted` or `filtered`) version of 
+   * an existing collection, so we should not create or remove listeners or clear the cache
+   * when this is torn down.
+   */
+  isDerived?: boolean;
+};
+
+/**
+ * Creates a proxy around a {@link Realm.Collection} that will create new {@link Realm.Object}
+ * references on any relevant change (update, insert, deletion) and return the same
+ * object reference if no changes have occurred since the last access.
+ *
+ * This makes the {@link Realm.Collection} behave in an immutable way, as React expects, so
+ * that a {@link Realm.Object} can be wrapped in {@link React.memo} to prevent unnecessary
+ * rendering (see `useCollection` hook).
+ *
+ * @param args {@link CachedCollectionArgs} object arguments
+ * @returns Proxy object wrapping the collection
+ */
+export function cachedCollection<T extends Realm.Object>({
+  collection,
+  updateCallback,
+  collectionCache = new Map(),
+  isFirst = true,
+}: {
+  collection: Realm.Collection<T>;
+  updateCallback: () => void;
+  collectionCache?: Map<string, T>;
```

Comment:
Semantically this is a cache of objects

```suggestion
  objectCache?: Map<string, T>;
```